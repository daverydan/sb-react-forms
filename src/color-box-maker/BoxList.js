import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
  const [boxes, setBoxes] = useState([
    { bgColor: "red", height: 50, width: 100 },
    { bgColor: "green", height: 100, width: 50 },
    { bgColor: "blue", height: 50, width: 50 },
  ]);

  const addBox = (bgColor, height, width) => {
    setBoxes((boxes) => [
      ...boxes,
      { bgColor, height: parseInt(height), width: parseInt(width) },
    ]);
  };

  const removeBox = (idx) => {
    const newBoxes = [...boxes];
    newBoxes.splice(idx, 1);
    setBoxes(newBoxes);
  };

  return (
    <>
      <NewBoxForm addBox={addBox} />
      <hr />
      <div className="BoxList" data-testid="box-list">
        {boxes.map((b, idx) => (
          <Box
            key={idx}
            id={idx}
            bgColor={b.bgColor}
            height={b.height}
            width={b.width}
            removeBox={() => removeBox(idx)}
          />
        ))}
      </div>
    </>
  );
};

export default BoxList;
