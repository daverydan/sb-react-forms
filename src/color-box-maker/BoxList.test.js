import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

test("app not broken", () => {
  render(<BoxList />);
});

test("it matches snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

test("should add new item", () => {
  const { queryByText, queryByLabelText, getByTestId } = render(<BoxList />);
  const boxList = getByTestId("box-list");
  const bgColor = queryByLabelText("Background Color:");
  const height = queryByLabelText("Height:");
  const width = queryByLabelText("Width:");
  const button = queryByText("Add");
  fireEvent.change(bgColor, { target: { value: "#F50000" } });
  fireEvent.change(height, { target: { value: 50 } });
  fireEvent.change(width, { target: { value: 50 } });
  fireEvent.click(button);
  const box3 = document.getElementById("box-3");
  expect(box3).toBeInTheDocument();
  expect(box3.style.backgroundColor).toBe("rgb(245, 0, 0)");
  expect(box3.style.height).toBe("50px");
  expect(box3.style.width).toBe("50px");
  expect(boxList.children.length).toBe(4);
});

test("should delete an item", () => {
  const { getByTestId } = render(<BoxList />);
  const boxList = getByTestId("box-list");
  const boxBtn2 = getByTestId("box-btn-2");
  fireEvent.click(boxBtn2);
  const box2 = document.getElementById("box-2");
  expect(box2).not.toBeInTheDocument();
  expect(boxList.children.length).toBe(2);
});
