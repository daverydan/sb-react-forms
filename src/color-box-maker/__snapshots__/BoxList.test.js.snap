// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="NewBoxForm"
  >
    <h1>
      Color Box Maker
    </h1>
    <form>
      <div
        class="control"
      >
        <label
          for="bgColor"
        >
          Background Color:
        </label>
        <input
          id="bgColor"
          name="bgColor"
          type="color"
          value="#F50000"
        />
      </div>
      <div
        class="control"
      >
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          type="number"
          value="50"
        />
      </div>
      <div
        class="control"
      >
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          type="number"
          value="50"
        />
      </div>
      <button>
        Add
      </button>
    </form>
  </div>
  <hr />
  <div
    class="BoxList"
    data-testid="box-list"
  >
    <div
      class="Box"
      id="box-0"
      style="background-color: red; height: 50px; width: 100px;"
    >
      <button
        data-testid="box-btn-0"
      >
        x
      </button>
    </div>
    <div
      class="Box"
      id="box-1"
      style="background-color: green; height: 100px; width: 50px;"
    >
      <button
        data-testid="box-btn-1"
      >
        x
      </button>
    </div>
    <div
      class="Box"
      id="box-2"
      style="background-color: blue; height: 50px; width: 50px;"
    >
      <button
        data-testid="box-btn-2"
      >
        x
      </button>
    </div>
  </div>
</DocumentFragment>
`;
